File: shared/src/commonMain/kotlin/club/djipi/lebarcs/shared/domain/logic/MoveValidator.kt
package club.djipi.lebarcs.shared.domain.logic

import club.djipi.lebarcs.shared.domain.model.Board
import club.djipi.lebarcs.shared.domain.model.BoardPosition

object MoveValidator {

    /**
     * Valide la géométrie d'un coup.
     * Vérifie que toutes les tuiles posées sont sur une seule ligne (horizontale ou verticale).
     * @param placedTiles Les positions des tuiles qui viennent d'être jouées.
     * @return true si le placement est valide, false sinon.
     */
    fun isPlacementValid(board: Board, placedTiles: Set<BoardPosition>): Boolean {
        if (placedTiles.size <= 1) {
            // Un coup d'une seule tuile est toujours aligné.
            // On vérifiera la connexion à une tuile existante ailleurs.
            return true
        }

        // 1. Vérifier que les NOUVELLES tuiles sont bien alignées
        val allOnSameRow = placedTiles.map { it.row }.toSet().size == 1
        val allOnSameCol = placedTiles.map { it.col }.toSet().size == 1

        if (!allOnSameRow && !allOnSameCol) {
            println("Validation échec : les tuiles posées ne sont pas sur la même ligne/colonne.")
            return false
        }

        // 2. Vérifier la contiguïté du mot principal formé
        if (allOnSameRow) {
            // Coup horizontal
            val row = placedTiles.first().row
            val minCol = placedTiles.minOf { it.col }
            val maxCol = placedTiles.maxOf { it.col }

            // On vérifie toutes les cases entre la première et la dernière NOUVELLE tuile
            for (col in minCol..maxCol) {
                // S'il y a un trou (case vide), c'est invalide
                if (board.getTile(BoardPosition(row, col)) == null && BoardPosition(row, col) !in placedTiles) {
                    println("Validation échec : trou dans le mot principal horizontal.")
                    return false
                }
            }
        } else { // Coup vertical
            val col = placedTiles.first().col
            val minRow = placedTiles.minOf { it.row }
            val maxRow = placedTiles.maxOf { it.row }

            for (row in minRow..maxRow) {
                if (board.getTile(BoardPosition(row, col)) == null && BoardPosition(row, col) !in placedTiles) {
                    println("Validation échec : trou dans le mot principal vertical.")
                    return false
                }
            }
        }

            println("Validation placement : OK !")
            return true
        }


fun isMoveConnected(board: Board, placedTiles: Set<BoardPosition>, turnNumber: Int): Boolean {
    // Règle du premier tour : le coup doit passer par le centre.
    if (turnNumber == 1) {
        val centerBoardPosition = BoardPosition(Board.SIZE / 2, Board.SIZE / 2) // (7,7)
        return centerBoardPosition in placedTiles
    }

    // Règle des tours suivants : au moins une tuile doit être adjacente à une tuile existante.
    val adjacentOffsets = listOf(BoardPosition(-1, 0), BoardPosition(1, 0), BoardPosition(0, -1), BoardPosition(0, 1))

    for (pos in placedTiles) {
        for (offset in adjacentOffsets) {
            val adjacentPos = BoardPosition(pos.row + offset.row, pos.col + offset.col)
            if (adjacentPos.row in 0 until Board.SIZE && adjacentPos.col in 0 until Board.SIZE) {
                // Si la case adjacente a une tuile ET qu'elle ne fait pas partie du coup actuel
                if (board.getTile(adjacentPos) != null && adjacentPos !in placedTiles) {
                    return true // C'est connecté !
                }
            }
        }
    }

    // Si on a vérifié toutes les tuiles et leurs voisins sans trouver de connexion...
    return false
}}

File: node_server/src/logic/MoveValidator.ts
import type { Board, BoardPosition } from '../models/GameModels.js';

const BOARD_SIZE = 15;

function getTile(board: Board, position: BoardPosition): any | null {
    // ... (copiez la même fonction getTile que dans WordFinder)
    if (position.row >= 0 && position.row < BOARD_SIZE && position.col >= 0 && position.col < BOARD_SIZE) {
        return board[position.row]![position.col]!.tile;
    }
    return null;
}

export function isPlacementValid(board: Board, placedTiles: BoardPosition[]): boolean {
    if (placedTiles.length <= 1) return true;

    const rows = new Set(placedTiles.map(p => p.row));
    const cols = new Set(placedTiles.map(p => p.col));

    const allOnSameRow = rows.size === 1;
    const allOnSameCol = cols.size === 1;

    if (!allOnSameRow && !allOnSameCol) {
        console.log("Validation échec : les tuiles posées ne sont pas sur la même ligne/colonne.");
        return false;
    }

    if (allOnSameRow) {
        const row = placedTiles[0]!.row;
        const minCol = Math.min(...placedTiles.map(p => p.col));
        const maxCol = Math.max(...placedTiles.map(p => p.col));

        for (let col = minCol; col <= maxCol; col++) {
            const currentPos = { row, col };
            const hasExistingTile = getTile(board, currentPos) !== null;
            const isNewTile = placedTiles.some(p => p.row === row && p.col === col);
            if (!hasExistingTile && !isNewTile) {
                console.log(`Validation échec : trou dans le mot principal horizontal.`);
                return false;
            }
        }
    } else { // All on same column
        const col = placedTiles[0]!.col;
        const minRow = Math.min(...placedTiles.map(p => p.row));
        const maxRow = Math.max(...placedTiles.map(p => p.row));

        for (let row = minRow; row <= maxRow; row++) {
            const currentPos = { row, col };
            const hasExistingTile = getTile(board, currentPos) !== null;
            const isNewTile = placedTiles.some(p => p.row === row && p.col === col);
            if (!hasExistingTile && !isNewTile) {
                console.log(`Validation échec : trou dans le mot principal vertical.`);
                return false;
            }
        }
    }

    return true;
}

export function isMoveConnected(board: Board, placedTiles: BoardPosition[], turnNumber: number): boolean {
    const centerPos = Math.floor(BOARD_SIZE / 2);

    if (turnNumber === 1) {
        return placedTiles.some(p => p.row === centerPos && p.col === centerPos);
    }

    const adjacentOffsets = [{row: -1, col: 0}, {row: 1, col: 0}, {row: 0, col: -1}, {row: 0, col: 1}];
    const placedSet = new Set(placedTiles.map(p => `${p.row},${p.col}`));

    for (const pos of placedTiles) {
        for (const offset of adjacentOffsets) {
            const adjacentPos = { row: pos.row + offset.row, col: pos.col + offset.col };
            if (getTile(board, adjacentPos) !== null && !placedSet.has(`${adjacentPos.row},${adjacentPos.col}`)) {
                return true; // C'est connecté !
            }
        }
    }

    return false;
}